synapse/handlers/auth.py

```diff
@@ -46,6 +46,13 @@ def __init__(self, hs):
         }
         self.bcrypt_rounds = hs.config.bcrypt_rounds
         self.sessions = {}
+        self.INVALID_TOKEN_HTTP_STATUS = 401
+        self._KNOWN_CAVEAT_PREFIXES = set([
```

Comment:
Why are we duplicating this (and basically all of the code below in this file) from synapse/api/auth.py rather than referring to it?

You can get an instance of synapse/api/auth.py by calling self.hs.get_auth() (you may need to self.hs = hs in this constructor).

From there, I'd fold all of the new macaroon functions into extra arguments on the existing functions, and call those.

It looks like you'll need to add an optional type parameter to _validate_macaroon, drop some leading underscores from function names, and then everything else should just fall out for free by calling functions on self.hs.get_auth()